File: plugins/ralph-universal/hooks/pretooluse-loop-guard.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = ["pydantic>=2.10.0", "filelock>=3.20.0"]
# ///
"""PreToolUse hook: Guard loop control files from deletion.

Prevents Claude from bypassing the Stop hook by directly running
Bash commands that delete .claude/loop-enabled or other loop files.

NOTE: Idle command detection (git status spam) is handled by the Stop hook
using stamina-style exponential backoff in loop-until-done.py.

Protected files and deletion patterns are configurable via
.claude/ralph-config.json.

ADR: /docs/adr/2025-12-20-ralph-rssi-eternal-loop.md
ADR: /docs/adr/2025-12-17-posttooluse-hook-visibility.md (output format)
"""

import json
import os
import re
import sys

from core.config_schema import ProtectionConfig, load_config

def get_protection_config() -> ProtectionConfig:
    """Get protection parameters from config."""
    project_dir = os.environ.get("CLAUDE_PROJECT_DIR", "")
    config = load_config(project_dir if project_dir else None)
    return config.protection


def is_official_ralph_command(command: str) -> bool:
    """Check if command is an official Ralph command with bypass marker.

    Any command containing a registered bypass marker (e.g., RALPH_STOP_SCRIPT,
    RALPH_ENCOURAGE_SCRIPT) is allowed to operate on protected files.
    """
    cfg = get_protection_config()
    # Check new bypass_markers list first
    for marker in cfg.bypass_markers:
        if marker in command:
            return True
    # Fallback to legacy single marker for backward compatibility
    return cfg.stop_script_marker in command


def is_deletion_command(command: str) -> bool:
    """Check if command attempts to delete protected files."""
    cfg = get_protection_config()

    # Check for deletion patterns (from config)
    has_deletion_cmd = any(
        re.search(pattern, command) for pattern in cfg.deletion_patterns
    )

    if not has_deletion_cmd:
        return False

    # Check if any protected file is mentioned (from config)
    for protected_file in cfg.protected_files:
        # Check for full path or relative path
        if protected_file in command:
            return True
        # Check for just the filename
        filename = os.path.basename(protected_file)
        if filename in command:
            return True

    return False


def main():
    """Check Bash command and block if it deletes loop files."""
    # ===== ALPHA-FORGE ONLY GUARD =====
    # Ralph is dedicated to alpha-forge ML research workflows only.
    # Skip all processing for non-alpha-forge projects (zero overhead).
    project_dir = os.environ.get("CLAUDE_PROJECT_DIR", "")
    if project_dir:
        from core.project_detection import is_alpha_forge_project
        if not is_alpha_forge_project(project_dir):
            # Silent pass-through: allow command, no Ralph processing
            # Using modern permissionDecision format (not deprecated decision:allow)
            print(json.dumps({
                "hookSpecificOutput": {
                    "hookEventName": "PreToolUse",
                    "permissionDecision": "allow"
                }
            }))
            return

    # Read tool input from stdin
    try:
        tool_input = json.load(sys.stdin)
    except json.JSONDecodeError as e:
        # Can't parse input, allow the command but warn
        print(f"[ralph] Warning: Failed to parse tool input: {e}", file=sys.stderr)
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": "allow"
            }
        }))
        return

    # Get the command being executed
    command = tool_input.get("command", "")

    # Helper for allow response (modern format)
    def allow_command():
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": "allow"
            }
        }))

    if not command:
        allow_command()
        return

    # Allow official Ralph commands to operate on protected files
    if is_official_ralph_command(command):
        allow_command()
        return

    # Check if this is a deletion attempt on protected files
    if is_deletion_command(command):
        # Using modern permissionDecision format (not deprecated decision:block)
        result = {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": "deny",
                "permissionDecisionReason": (
                    "[RALPH LOOP GUARD] Cannot delete loop control files. "
                    "The Ralph autonomous loop is active. Only the user can stop it "
                    "by running /ralph:stop or removing .claude/loop-enabled manually. "
                    "Continue working on improvement opportunities instead."
                ),
            }
        }
        print(json.dumps(result))
        return

    # Allow all other commands
    allow_command()


if __name__ == "__main__":
    main()

File: plugins/ru/hooks/pretooluse-loop-guard.ts
#!/usr/bin/env bun
/**
 * PreToolUse hook: Guard loop control files from deletion.
 *
 * Prevents Claude from bypassing the Stop hook by directly running
 * Bash commands that delete .claude/loop-enabled or other loop files.
 *
 * Protected files and deletion patterns are configurable via
 * .claude/ru-config.json.
 *
 * ADR: /docs/adr/2025-12-20-ralph-rssi-eternal-loop.md
 * ADR: /docs/adr/2025-12-17-posttooluse-hook-visibility.md (output format)
 */

import { basename } from "path";
import { loadConfig, type ProtectionConfig } from "./core/config-schema";
import { trackHookError } from "../../itp-hooks/hooks/lib/hook-error-tracker.ts";

// --- Types ---

interface HookInput {
  tool_name: string;
  tool_input: {
    command?: string;
  };
}

interface PreToolUseOutput {
  hookSpecificOutput: {
    hookEventName: "PreToolUse";
    permissionDecision: "allow" | "deny";
    permissionDecisionReason?: string;
  };
}

// --- Helper Functions ---

function getProtectionConfig(): ProtectionConfig {
  const projectDir = process.env.CLAUDE_PROJECT_DIR || "";
  const config = loadConfig(projectDir || undefined);
  return config.protection;
}

function allowCommand(): void {
  const output: PreToolUseOutput = {
    hookSpecificOutput: {
      hookEventName: "PreToolUse",
      permissionDecision: "allow",
    },
  };
  console.log(JSON.stringify(output));
}

function denyCommand(reason: string): void {
  const output: PreToolUseOutput = {
    hookSpecificOutput: {
      hookEventName: "PreToolUse",
      permissionDecision: "deny",
      permissionDecisionReason: reason,
    },
  };
  console.log(JSON.stringify(output));
}

/**
 * Check if command is an official Ralph command with bypass marker.
 *
 * Any command containing a registered bypass marker (e.g., RALPH_STOP_SCRIPT,
 * RALPH_ENCOURAGE_SCRIPT) is allowed to operate on protected files.
 */
function isOfficialRalphCommand(command: string): boolean {
  const cfg = getProtectionConfig();

  // Check bypass_markers list
  for (const marker of cfg.bypass_markers) {
    if (command.includes(marker)) {
      return true;
    }
  }

  // Fallback to legacy single marker for backward compatibility
  return command.includes(cfg.stop_script_marker);
}

/**
 * Check if command attempts to delete protected files.
 */
function isDeletionCommand(command: string): boolean {
  const cfg = getProtectionConfig();

  // Check for deletion patterns (from config)
  const hasDeletionCmd = cfg.deletion_patterns.some((pattern) => {
    try {
      return new RegExp(pattern).test(command);
    } catch {
      // Invalid regex pattern, skip it
      return false;
    }
  });

  if (!hasDeletionCmd) {
    return false;
  }

  // Check if any protected file is mentioned (from config)
  for (const protectedFile of cfg.protected_files) {
    // Check for full path or relative path
    if (command.includes(protectedFile)) {
      return true;
    }
    // Check for just the filename
    const filename = basename(protectedFile);
    if (command.includes(filename)) {
      return true;
    }
  }

  return false;
}

// --- Main ---

async function main(): Promise<void> {
  // Read tool input from stdin
  let inputText = "";
  for await (const chunk of Bun.stdin.stream()) {
    inputText += new TextDecoder().decode(chunk);
  }

  let toolInput: HookInput;
  try {
    toolInput = JSON.parse(inputText);
  } catch (e) {
    // Can't parse input, allow the command but warn
    trackHookError("pretooluse-loop-guard", `Failed to parse tool input: ${e}`);
    allowCommand();
    return;
  }

  const command = toolInput.tool_input?.command || "";

  if (!command) {
    allowCommand();
    return;
  }

  // Allow official Ralph commands to operate on protected files
  if (isOfficialRalphCommand(command)) {
    allowCommand();
    return;
  }

  // Check if this is a deletion attempt on protected files
  if (isDeletionCommand(command)) {
    denyCommand(
      "[RALPH LOOP GUARD] Cannot delete loop control files. " +
        "The Ralph autonomous loop is active. Only the user can stop it " +
        "by running /ru:stop or removing .claude/loop-enabled manually. " +
        "Continue working on improvement opportunities instead."
    );
    return;
  }

  // Allow all other commands
  allowCommand();
}

main().catch((e) => {
  trackHookError("pretooluse-loop-guard", `Error in pretooluse-loop-guard: ${e}`);
  // On error, allow command to avoid blocking legitimate operations
  allowCommand();
});
